cli/pkg/kctrl/cmd/package/release/helm_openapi_schema_gen.go

```diff
@@ -0,0 +1,280 @@
+		return &kcdatav1alpha1.ValuesSchema{
+			OpenAPIv3: runtime.RawExtension{Raw: nil},
+		}, nil
+	}
+
+	if document.Kind != yaml3.DocumentNode {
+		return nil, fmt.Errorf("invalid node kind supplied: %d", document.Kind)
+	}
+	if document.Content[0].Kind != yaml3.MappingNode {
+		return nil, fmt.Errorf("values file must resolve to a map (was %d)", document.Content[0].Kind)
+	}
+
+	openAPIProperties, err := h.calculateProperties(nil, document.Content[0])
+	if err != nil {
+		return nil, err
+	}
+
+	bytes, err := yaml2.Marshal(h.toYAML(openAPIProperties))
+	if err != nil {
+		return nil, err
+	}
+	jsonEncodedBytes, err := yaml.YAMLToJSON(bytes)
+	if err != nil {
+		return nil, err
+	}
+	return &kcdatav1alpha1.ValuesSchema{
+		OpenAPIv3: runtime.RawExtension{Raw: jsonEncodedBytes},
+	}, nil
+}
+
+func (h HelmValuesSchemaGen) readValuesFile() ([]byte, error) {
+	fileInfo, err := os.Stat(h.dirPath)
+	if err != nil {
+		return nil, err
+	}
+	if !fileInfo.IsDir() {
+		return nil, fmt.Errorf("expected %s to be directory", h.dirPath)
+	}
+
+	fileData, err := os.ReadFile(filepath.Join(h.dirPath, "values.yaml"))
+	if err != nil {
+		// It is possible that helm chart doesn't have values.yml file.
+		if os.IsNotExist(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+	return fileData, nil
+}
+
+func (h HelmValuesSchemaGen) toYAML(val interface{}) interface{} {
+	switch typedVal := val.(type) {
+	case *Map:
+		result := yaml2.MapSlice{}
+		typedVal.Iterate(func(k string, v interface{}) {
+			result = append(result, yaml2.MapItem{
+				Key:   k,
+				Value: h.toYAML(v),
+			})
+		})
+		return result
+	default:
+		return val
+	}
+}
+
+func (h HelmValuesSchemaGen) calculateProperties(key *yaml3.Node, value *yaml3.Node) (*Map, error) {
+	var apiKeys []*MapItem
+	description, isPresent := h.getDescriptionFromNode(key)
+	if isPresent {
+		apiKeys = append(apiKeys, &MapItem{Key: descriptionKey, Value: description})
+	}
+
+	switch value.Kind {
+	case yaml3.MappingNode:
+		var properties []*MapItem
+		apiKeys = append(apiKeys, &MapItem{Key: typeKey, Value: objectVal})
+
+		for i := 0; i < len(value.Content); i += 2 {
+			calculatedProperties, err := h.calculateProperties(value.Content[i], value.Content[i+1])
+			if err != nil {
+				return nil, err
+			}
+			properties = append(properties, &MapItem{Key: calculatedProperties.Items[0].Key, Value: calculatedProperties.Items[0].Value})
+		}
+		if len(properties) > 0 {
+			apiKeys = append(apiKeys, &MapItem{Key: propertiesKey, Value: &Map{Items: properties}})
+		} else {
+			apiKeys = append(apiKeys, &MapItem{Key: defaultKey, Value: "{}"})
+		}
+	case yaml3.SequenceNode:
+		apiKeys = append(apiKeys, &MapItem{Key: typeKey, Value: arrayVal})
+		apiKeys = append(apiKeys, &MapItem{Key: defaultKey, Value: []interface{}{}})
+
+		if len(value.Content) > 0 {
+			// For now, we have decided to add only the first item if it is an array.
+			// Based on feedback, we will revisit and enhance the implementation.
```

Comment:
nit: Lets keep the comment concise,
`TODO: Do we need to consider that elements in a list might have different or more keys?`